grai-server/app/installations/github.py

```diff
@@ -0,0 +1,128 @@
+            signing_key = jwt.jwk_from_pem(pem_file.read())
+
+        payload = {"iat": int(time.time()), "exp": int(time.time()) + 600, "iss": app_id}
+
+        jwt_instance = jwt.JWT()
+        encoded_jwt = jwt_instance.encode(payload, signing_key, alg="RS256")
+
+        return encoded_jwt
+
+    def connect(self):
+        jwt = self.generate_jwt()
+
+        res = requests.post(
+            f"https://api.github.com/app/installations/{self.installation_id}/access_tokens",
+            headers={
+                "Authorization": f"Bearer {jwt}",
+                "Accept": "application/vnd.github+json",
+            },
+        )
+
+        res.raise_for_status()
+
+        data = res.json()
+
+        print(f"data: {data}")
+
+        self.token = data.get("token")
+        self.expires_at = data.get("expires_at")
+
+        print(f"token: {self.token}")
+
+        if self.token is None:
+            raise Exception(data.get("message"))
+
+        self.api = GhApi(owner=self.owner, repo=self.repo, token=self.token)
+
+    def get_api(self) -> GhApi:
+        if not self.api:
+            self.connect()
+
+        return self.api
+
+    def create_check(
+        self, head_sha: str, external_id: str = None, name: str = "Grai Update", details_url: str = None, output=None
+    ):
+        check = self.api.checks.create(
+            name=name,
+            head_sha=head_sha,
+            external_id=external_id,
+            status="queued",
+            details_url=details_url,
+            output=output,
+        )
+
+        return check
+
+    def start_check(self, check_id: int):
+        return self.api.checks.update(
+            check_run_id=check_id, status="in_progress", started_at=datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
+        )
+
+    def complete_check(self, check_id: int, conclusion: str = "success"):
+        return self.api.checks.update(
+            check_run_id=check_id,
+            status="completed",
+            conclusion=conclusion,
+            completed_at=datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ"),
+        )
+
+    def get_repos(self):
+        return self.api.apps.list_repos_accessible_to_installation()["repositories"]
+
+    @staticmethod
+    def add_comment_identifier(message, identifier):
+        return f"{identifier}{message}"
+
+    def get_marked_comment(self, pr_number: str, identifier: str) -> Optional[dict]:
+        current_comments = self.api.issues.list_comments(pr_number)
+        # user_comments = (comment for comment in current_comments if comment["user"]["id"] == self.bot_user_id)
+        for comment in current_comments:
+            if identifier in comment["body"]:
+                return comment
+
+        return None
+
+    def post_comment(self, pr_number: str, message: str):
+        message = self.add_comment_identifier(message, self.test_signal_text)
+
+        marked_comment = self.get_marked_comment(pr_number, self.test_signal_text)
+
+        if marked_comment is None:
```

Comment:
I have a PR up to delete comments if all tests are fixed we can use here.
https://github.com/grai-io/grai-actions/pull/14/files

Reply:
My feeling is that we should comment if all the tests pass, so that the user can see that, however happy to debate.